src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -369,5 +403,14 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
 
             return restoreContext;
         }
+
+        private class ProjectRestoreSpec
+        {
+            public DependencyGraphSpec DgSpec { get; set; }
+            public IReadOnlyList<PackageSpec> PackageSpecs { get; set; }
+            public IReadOnlyList<IAssetsLogMessage> ProjectAdditionalMessages { get; set; }
+            public HashSet<string> KnownProjects { get; set; }
```

Comment:
This is used in a non thread safe isn't it?

Reply:
Made it thread safe by using`ConcurrentDictionary`, please check.

Reply:
Also changed `AllAdditionalMessages` to `ConcurrentBag` type under assumption of order of element in it doesn't matter.